compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2254,26 +2278,40 @@ class Typer extends Namer
     trees mapconserve (typed(_))
 
   def typedStats(stats: List[untpd.Tree], exprOwner: Symbol)(implicit ctx: Context): List[Tree] = {
+    val (stats1, _) = typedStatsAndGetFacts(stats, exprOwner)
+    stats1
+  }
+
+  def typedStatsAndGetFacts(stats: List[untpd.Tree], exprOwner: Symbol)(implicit ctx: Context): (List[Tree], FlowFacts) = {
     val buf = new mutable.ListBuffer[Tree]
     val enumContexts = new mutable.HashMap[Symbol, Context]
       // A map from `enum` symbols to the contexts enclosing their definitions
-    @tailrec def traverse(stats: List[untpd.Tree])(implicit ctx: Context): List[Tree] = stats match {
+    @tailrec def traverse(stats: List[untpd.Tree], facts: FlowFacts)(implicit ctx: Context): (List[Tree], FlowFacts) = stats match {
```

Comment:
It's very confusing that there is both an explicitly maintained `facts` and a context that maintains flow facts. There _must_ be a simpler way! This is getting put of hand. `typedStats` was hard to understand and maintain before. Now it becomes impossible... Let's take a step back and see whether we can find a better way to handle flow facts.